### PR TITLE
#issue 114 Rate Limiting 

### PR DESCRIPTION
**Pull request by anandrgit**
Rate Limit  using ( token bucket algorithm)
Have the Burst size and the events per min defined in the config
Rate Limit the requests based on these parameters.
Have a sequence of interceptors passed to the grpc server.
All requests will first pass through rate limit interceptor and only then pass to the Payment Validation interceptor

### CODE REVIEW

**Comment by anandrgit** [comment]
@vsbogd  Request your review

File: config/config.go [hunk 1]
```diff
@@ -52,6 +55,7 @@ const (
 	"ipfs_end_point": "http://localhost:5002/", 
 	"organization_name": "ExampleOrganization", 
 	"passthrough_enabled": false,
+	"rate_limit": 20,
```
Comment (vsbogd) [inline config/config.go]: Is 20 request per minute enough? I would set some bigger value to not limiting users by default, may be even unlimited rate.

File: snetd/cmd/components.go [hunk 2]
```diff
@@ -190,12 +191,12 @@ func (components *Components) GrpcInterceptor() grpc.StreamServerInterceptor {
...
-		components.grpcInterceptor = handler.NoOpInterceptor
+		components.grpcInterceptor = grpc_middleware.ChainStreamServer(handler.GrpcRateLimitInterceptor(), handler.NoOpInterceptor)
 	} else {
 		log.Info("Blockchain is enabled: instantiate payment validation interceptor")
-		components.grpcInterceptor = handler.GrpcStreamInterceptor(
-			components.EscrowPaymentHandler(),
-		)
+		components.grpcInterceptor =
+			grpc_middleware.ChainStreamServer(handler.GrpcRateLimitInterceptor(), handler.GrpcStreamInterceptor(
+				components.EscrowPaymentHandler()))
```
Comment (vsbogd) [inline snetd/cmd/components.go]: Minor and more about code duplication and readability: you can introduce new component ```GrpcPaymentValidationInterceptor``` and move ```if !components.Blockchain().Enabled()``` condition into component initialization procedure, then you will have ```GrpcPaymentValidationInterceptor``` component which is renamed version of old ```GrpcInterceptor```:
```golang
 func (components *Components) GrpcPaymentValidationInterceptor() grpc.StreamServerInterceptor {
 	if components.grpcPaymentValidationInterceptor != nil {
 		return components.grpcPaymentValidationInterceptor
 	}
	if !components.Blockchain().Enabled() {
		log.Info("Blockchain is disabled: no payment validation")
		components.grpcPaymentValidationInterceptor = handler.NoOpInterceptor
	} else {
		log.Info("Blockchain is enabled: instantiate payment validation interceptor")
		components.grpcPaymentValidationInterceptor = handler.GrpcPaymentValidationInterceptor(components.EscrowPaymentHandler())
        }
	return components.grpcPaymentValidationInterceptor
}
```
And new ```GrpcInterceptor``` which chains other interceptor:
```golang
 func (components *Components) GrpcInterceptor() grpc.StreamServerInterceptor {
 	if components.grpcInterceptor != nil {
 		return components.grpcInterceptor
 	}
	components.grpcInterceptor = grpc_middleware.ChainStreamServer(handler.GrpcRateLimitInterceptor(), components.GrpcPaymentValidationInterceptor())
	return components.grpcInterceptor
}
```

**Comment by vsbogd** [comment]
I have pressed 'Requesting changes' by accident but in fact I don't see any bugs in code. All of the comments I made are more about readability and usability.

File: config/config.go (hunk 1 above)
Comment (anandrgit) [reply config/config.go]: Thansk @vsbogd , then lets remove the  "rate_limit": 20 in the json , we default to infinity in the code and this property becomes optional in its true sense. Burst size will have no meaning when the rate is infinity , we will explain this in the documentation at README.md

Comment (anandrgit) [reply config/config.go]: lets also remove the burst_size and default it to math.MaxInt64, from the config json.

File: config/config.go [hunk 3]
```diff
@@ -33,6 +34,7 @@ const (
 	PassthroughEnabledKey          = "passthrough_enabled"
 	PassthroughEndpointKey         = "passthrough_endpoint"
 	PrivateKeyKey                  = "private_key"
+	RateLimitPerMinute             = "rate_limit"
```
[the remark this answers is not in the document]
Comment (anandrgit) [reply config/config.go]: Sure, will make this change

File: ratelimit/rateLimit.go [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+
+func GetRateLimiter() rate.Limiter {
+	limiter := rate.NewLimiter(getLimit(), config.GetInt(config.BurstSize)) //separate config for Burst size ? or derive this from ratelimit ( say twice the rate limit) ?
+	return *limiter
+}
+
+func getLimit() rate.Limit {
+	configRateLimit := config.GetInt(config.RateLimitPerMinute)
+	ratePerMin := float64(configRateLimit)
+	intervalMsec := time.Duration((1/ratePerMin)*60*1000) * time.Millisecond
```
[the remark this answers is not in the document]
Comment (anandrgit) [reply ratelimit/rateLimit.go]: Sure

File: ratelimit/rateLimit.go [hunk 5]
```diff
@@ -0,0 +1,19 @@
+package ratelimit
+
+import (
+	"github.com/singnet/snet-daemon/config"
+	"golang.org/x/time/rate"
+	"time"
+)
+
+func GetRateLimiter() rate.Limiter {
```
[the remark this answers is not in the document]
Comment (anandrgit) [reply ratelimit/rateLimit.go]: Makes sense, will do

File: snetd/cmd/components.go (hunk 2 above)
Comment (anandrgit) [reply snetd/cmd/components.go]: Makes Sense, Fixed as a separate commit

File: handler/interceptors.go [hunk 6]
```diff
@@ -96,6 +98,30 @@ type PaymentHandler interface {
...
+	if e != nil {
+		log.WithError(e)
+		return e
+	}
+	return nil
+}
+
 // GrpcStreamInterceptor returns gRPC interceptor to validate payment. If
 // blockchain is disabled then noOpInterceptor is returned.
 func GrpcStreamInterceptor(defaultPaymentHandler PaymentHandler, paymentHandler ...PaymentHandler) grpc.StreamServerInterceptor {
```
[the remark this answers is not in the document]
Comment (anandrgit) [reply handler/interceptors.go]: Makes Sense, Fixed as a separate commit

File: ratelimit/rateLimit.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (anandrgit) [reply ratelimit/rateLimit.go]: Apologies  , fixed this now

File: handler/interceptors.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (anandrgit) [reply handler/interceptors.go]: Apologies  , fixed this now

**Pull request merged**